Userland/Libraries/LibLocate/LocateDb.cpp

```diff
@@ -0,0 +1,220 @@
+        chown(locate_db_path, 0, 18);
+    }
+
+    m_file_handle = fopen(path.characters(), (mode == LocateDbMode::Write ? "wb" : "rb"));
+    fseek(m_file_handle, 0, SEEK_END);
+    m_file_size = ftell(m_file_handle);
+    fseek(m_file_handle, 0, SEEK_SET);
+}
+
+LocateDb::~LocateDb()
+{
+    fclose(m_file_handle);
+}
+
+void LocateDb::write_header()
+{
+    // Format: magic string + version.
+    String db_header = String::formatted("{}|{}", "locatedb", 1);
+    fwrite(db_header.characters(), sizeof(uint8_t), db_header.length() + 1, m_file_handle);
+}
+
+void LocateDb::write_directory(DirectoryInfo& dir_info)
+{
+    struct DirStartEndHeader dir_header {
+        .type = (uint8_t)ChunkType::DirStart,
+        .db_id = dir_info.db_id,
+        .parent_db_id = dir_info.parent_db_id,
+        .path_size = dir_info.path.length() + 1
+    };
+    fwrite(bit_cast<char*>(&dir_header), sizeof(struct DirStartEndHeader), 1, m_file_handle);
+    fwrite(dir_info.path.characters(), sizeof(uint8_t), dir_info.path.length() + 1, m_file_handle);
+
+    for (auto& child : dir_info.children) {
+        ChildFileData cur_child {
+            .type = (uint8_t)(child->type == FileType::File ? ChunkType::File : ChunkType::Dir),
+            .db_id = child->db_id,
+            .parent_db_id = child->parent_db_id,
+            .user_owner = child->user_owner,
+            .group_owner = child->group_owner,
+            .st_mode = child->st_mode,
+            .name_size = child->name.length() + 1
+        };
+
+        fwrite(bit_cast<char*>(&cur_child), sizeof(struct ChildFileData), 1, m_file_handle);
+        fwrite(child->name.characters(), sizeof(uint8_t), child->name.length() + 1, m_file_handle);
+    }
+
+    dir_header.type = (uint8_t)ChunkType::DirEnd;
+    fwrite(bit_cast<char*>(&dir_header), sizeof(struct DirStartEndHeader), 1, m_file_handle);
+    fwrite(dir_info.path.characters(), sizeof(uint8_t), dir_info.path.length() + 1, m_file_handle);
+}
+
+bool LocateDb::verify_header()
+{
+    char buffer[BUFSIZ];
+    size_t bytes_read = fread(&buffer, sizeof(char), strlen("locatedb|1") + 1, m_file_handle);
+    if (bytes_read > 0 && "locatedb|1"sv == buffer)
+        return true;
+    return false;
+}
+
+LocateDb::ChunkType LocateDb::identify_chunk()
+{
+    uint8_t type_byte;
+    size_t bytes_read = fread(&type_byte, sizeof(uint8_t), 1, m_file_handle);
+    fseek(m_file_handle, -1, SEEK_CUR);
+    VERIFY(feof(m_file_handle) == 0);
+    VERIFY(bytes_read == 1);
+    ChunkType found_type;
+    switch (ChunkType(type_byte)) {
+    case (ChunkType::DirStart):
+        found_type = ChunkType::DirStart;
+        break;
+    case (ChunkType::DirEnd):
+        found_type = ChunkType::DirEnd;
+        break;
+    case (ChunkType::Dir):
+        found_type = ChunkType::Dir;
+        break;
+    case (ChunkType::File):
+        found_type = ChunkType::File;
+        break;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+
+    return found_type;
```

Comment:
What's the reason for this to be a switch/case? Isn't this just a lengthy way of doing

```suggestion
    VERIFY(type_byte < ChunkType::__Count);
    return type_byte;
```

Reply:
Yeah, I clearly overenginereed that one now when you mention it... your solution is way more elegant :)